### PR TITLE
esthri-cli: serve: add support for configuring a list of allowed prefixes

### PR DESCRIPTION
**Pull request by silverjam**
Adds a list "directory" prefixes to allow, for example:

```
cargo run -- serve --bucket esthri-test --allowed-prefix test_upload/ --allowed-prefix test_list_directory/
```

### CODE REVIEW

File: crates/esthri-cli/src/main.rs [hunk 1]
```diff
@@ -231,13 +233,12 @@ fn call_real_aws() {
         .args(args)
         .stdout(Stdio::inherit())
         .status()
-        .expect(
-            format!(
+        .unwrap_or_else(|_| {
```
Comment (silverjam) [inline crates/esthri-cli/src/main.rs:236]: clippy fix-up

**Comment by john-michaelburke** [comment]
This is great. I think a useful addition, maybe in another PR, to this would be a tailored landing page if a user reaches a non-whitelisted path. Not sure if that would be something best handled here or by the esthri consumer.

File: crates/esthri-cli/src/http_server.rs [hunk 2]
```diff
@@ -224,10 +231,12 @@ fn with_s3_client(
...
     bucket: &str,
+    allowed_prefixes: &[String],
 ) -> impl Filter<Extract = (http::Response<Body>,), Error = warp::Rejection> + Clone {
     warp::path::full()
         .and(with_s3_client(s3_client))
         .and(with_bucket(bucket.to_owned()))
+        .and(with_allowed_prefixes(allowed_prefixes))
         .and(warp::query::<DownloadParams>())
         .and(warp::header::optional::<String>("if-none-match"))
         .and_then(download)
```
Comment (notoriaga) [inline crates/esthri-cli/src/http_server.rs:242]: Is it possible to do something like 

```rust
let path = warp::path::full()
    .and(with_s3_client(s3_client))        
    .and(with_bucket(bucket.to_owned()))
    .and(with_allowed_prefixes(allowed_prefixes))
    .and(warp::query::<DownloadParams>())
    .and(warp::query::<DownloadParams>())
    .and(warp::header::optional::<String>("if-none-match"))
    .and(warp::header::optional::<String>("if-none-match"))
    .and_then(download)
    .and_then(download)
if !allowed_prefixes.is_empty() {
    path.and(with_allowed_prefixes(allowed_prefixes));
} else {
    path
}
```
then the filter doesn't have to worry about an empty prefix list

Comment (silverjam) [reply crates/esthri-cli/src/http_server.rs:242]: First thought it probably not?  This creates a function signature, so if we conditionally add it, then it would produce two diferent function signatures.

Comment (silverjam) [reply crates/esthri-cli/src/http_server.rs:259]: I was able to do some preprocessing in the filter setup, which makes the code a little clearer

File: crates/esthri-cli/tests/integration/http_server.rs [hunk 3]
```diff
@@ -40,6 +41,88 @@ async fn test_fetch_object() {
     assert_eq!(body, "this file has contents\n");
 }
 
+#[tokio::test]
```
Comment (achubbic-snav) [inline crates/esthri-cli/tests/integration/http_server.rs:44]: Please add comment describing what the test aims to validate, including whether this addresses nominal behavior or edge cases if applicable.

**Review by achubbic-snav — APPROVED** [review]
Marking as approved and trusting @notoriaga 's vigilant eye- current knowledge of filters insufficient.

File: crates/esthri-cli/tests/integration/http_server.rs (hunk 3 above)
Comment (silverjam) [reply crates/esthri-cli/tests/integration/http_server.rs:44]: done
